### ISSUE TITLE
Entity storage optimization

### ISSUE THREAD

**Issue by fordN** [graphprotocol/graph-node#142]

**Comment by lutter** [comment]
Tasks:
[ ] - see if table partitioning can be used for this
[ ] - if that's not possible, look at using pg schemas (one schema per subgraph)

**Comment by lutter** [comment]
It won't be easy to use partitioning to split up the entities table with Postgres 9.6; the support for partitioning in that version is too bare-bones to do that effectively. But in experimenting with other approaches, I ran across using a GIN index for an entire subgraph. The basic idea is that for each subgraph we would create an index

```
create index asubgraph_idx on entities
 using gin(data jsonb_path_ops) where subgraph 'SUBGRAPH_ID'
```

This is based on some experimentation with a few queries. [This gist](https://gist.github.com/lutter/a868df05f579868df85a0c632ab6e026) contains details. I'd appreciate any suggestions for what other queries to measure before implementing this change.

The TL;DR of all this is that we would only need one index per subgraph, rather than one per (subgraph,entity,attribute) triple, greatly reducing the number of indexes on the `entities` table, and therefore reducing the number of locks queries against that table need.

Besides that, there are a few other low-hanging fruits that we should work on as part of this ticket. In particular, we should never query an entity for `data -> 'id' ->> 'data'`, but rather make sure we use the `id` column. Similarly, currently when we upsert an entity, we load the entity first from the db, merge the entity, and write the merged entity back. We should just have the database do the merge.

**Comment by leodasvacas** [comment]
@lutter Sounds really promising! It'd be good to also bench inequalities with numbers and containment in arrays.

**Comment by lutter** [comment]
@leodasvacas I looked at numeric inequalities. The two queries I played with were this for ENS:
```sql
-- ENS Domain, only 6 out of 280k rows have non-null ttl. Query matches
-- 4 of them
select * from entities
 where subgraph = 'QmNzRGD2f9AhFRYFJftQXkojx1EW4AQTtspcbHuTw6v69R'
   and entity='Domain'
   and (data->'ttl'->>'data')::int > 267251968;
-- takes 114ms
```

```sql
-- Livepeer Reward; each of the 3635 Reward entities has a rewardtoken value; the query
-- matches 1327 rows
select * from entities
 where subgraph='QmfPTP8Kao6kTGRRRDyHKnsQWgDU1ytAqhezJuP3Spq86A'
     and entity='Reward'
     and (data -> 'rewardTokens' ->> 'data')::numeric > 5e20;
-- takes 19ms
```

The interesting thing here is that both these queries do not use an attribute index; they instead use an index scan on `subgraph_entity_idx`. What I am proposing will therefore not affect query performance for these types of queries. What the planner is doing would be a full table scan on the `rewards` table if we had a relational schema for each entity in a subgraph. As the first query shows, that's fine for even our biggest 'real' subgraph.

The queries I was playing with were against CryptoKitties which is by far our largest subgraph (4.3M entities), followed by `michaelhly/zero_ex_events` (800k entities), 10 more subgraphs with 100k-300k entities, and all the rest has less than 100k entities.

**Comment by lutter** [comment]
One more note on our current attribute indexes: after the cleanup that @Jannis and I did this morning, we have 1469 attribute indexes; only 189 of them have actually ever been used in a query. We also saw a significant drop in CPU load on the DB and explorer speedup after we dropped ~ 1800 unneeded indexes.

It might therefore be reasonable to help with our performance woes with the following three steps:
1. Make sure we never query by id using `data->'id'->>'data'` and always use the `id` column.
2. Drop all attribute indexes
3. Add the GIN indexes that I have been describing

Steps 1 and 2 can be done pretty quickly, and after our experience this morning are likely to bring a huge performance boost. We are basically overindexing the entitites table, to the detriment of the overall performance of the system.

**Comment by Jannis** [comment]
@lutter The tasks above look good to me. Do we deem this sufficient to optimize how we store entities or is splitting up the entities table still part of this.

If not, we should probably rename this issue to `Optimize entity indexes` or similar.

**Comment by lutter** [comment]
I still think we should split up the `entities` table; I believe it will make the system much more resilient to growth in the number of subgraphs. Not sure if that should happen as part of this ticket. I feel this ticket has higher priority than splitting up the table, as getting rid of the large number of indexes that way will be quicker and reduce CPU load on the db.

**Comment by yanivtal** [comment]
Closing as a duplicate as #793
